neuraxle/metaopt/auto_ml.py

```diff
@@ -63,38 +68,76 @@ def load_all_trials(self) -> Tuple[List[HyperparameterSamples], List[float]]:
         pass
 
     @abstractmethod
-    def save_score_for_trial(self, hyperparams: HyperparameterSamples, score: float):
+    def save_score_for_success_trial(self, hyperparams: HyperparameterSamples, score: float):
         """
-        Save hyperparams, and score for a trial.
+        Save hyperparams, and score for a successful trial.
 
         :return: (hyperparams, scores)
         """
         pass
 
-    def get_best_hyperparams(self, higher_score_is_better=True) -> HyperparameterSamples:
+    @abstractmethod
+    def save_failure_for_trial(self, hyperparams: HyperparameterSamples, exception: Exception):
         """
-        Get the best hyperparams from all previous trials.
+        Save hyperparams, and score for a failed trial.
 
-        :return: best hyperparams
-        :rtype: HyperparameterSamples
+        :return: (hyperparams, scores)
         """
-        hyperparams, scores = self.load_all_trials()
-
-        best_score = None
-        best_hyperparams = None
-
-        for trial_hyperparam, trial_score in zip(hyperparams, scores):
-            if best_score is None or higher_score_is_better == (trial_score > best_score):
-                best_score = trial_score
-                best_hyperparams = trial_hyperparam
-
-        return best_hyperparams
+        pass
 
     def _get_trial_hash(self, hp_dict):
         current_hyperparameters_hash = hashlib.md5(str.encode(str(hp_dict))).hexdigest()
         return current_hyperparameters_hash
 
 
+class InMemoryHyperparamsRepository(HyperparamsRepository):
+    """
+    In memory hyperparams repository that can print information about trials.
+    Useful for debugging.
+
+    Example usage :
+
+    .. code-block:: python
+
+        InMemoryHyperparamsRepository(
+            print_new_trial=True,
+            print_success_trial=True,
+            print_exception=True
+        )
+
+
+    .. seealso::
+        :class:`HyperparamsRepository`,
+        :class:`HyperparameterSamples`,
+        :class:`AutoMLSequentialWrapper`
+    """
+
+    def __init__(self, print_new_trial=True, print_success_trial=True, print_exception=True):
```

Comment:
TODO later: 
Add an argument: `print_func: Callable = print`. This way people may be able to pass any logger function in the future. 
(in python 2, print was a special word, now it is truly a function).